File: take_u_forward/graph/20_Prerequisite Tasks.java
// User function Template for Java

class Solution {
    public boolean isPossible(int N, int P, int[][] prerequisites) {
        // Your Code goes here
         // Create adjacency list
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            adj.add(new ArrayList<>());
        }

        // Correct edge direction: v -> u
        for (int i = 0; i < P; i++) {
            int u = prerequisites[i][0];
            int v = prerequisites[i][1];
            adj.get(v).add(u);
        }
        
         // Indegree array
        int[] indegree = new int[N];
        for (int i = 0; i < N; i++) {
            for (int it : adj.get(i)) {
                indegree[it]++;
            }
        }

        // Queue for nodes with indegree 0
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < N; i++) {
            if (indegree[i] == 0) {
                q.add(i);
            }
        }

       int cnt=0;

        // Kahn's Algorithm
        while (!q.isEmpty()) {
            int node = q.poll();
            cnt++; 
            for (int it : adj.get(node)) {
                indegree[it]--;
                if (indegree[it] == 0) {
                    q.add(it);
                }
            }
        }

        if(cnt==N) return true;
        return false;
    }
}

File: take_u_forward/graph/20_Prerequisite Tasks.cpp
class Solution {
  public:
    bool isPossible(int N, int P, vector<pair<int, int> >& prerequisites) {
        // Code here
        // Step 1: Create adjacency list
        vector<vector<int>> adj(N);
        for (auto &p : prerequisites) {
            int u = p.first;
            int v = p.second;
            adj[v].push_back(u);   // v -> u
        }

        // Step 2: Calculate indegree
        vector<int> indegree(N, 0);
        for (int i = 0; i < N; i++) {
            for (int it : adj[i]) {
                indegree[it]++;
            }
        }

        // Step 3: Push nodes with indegree 0
        queue<int> q;
        for (int i = 0; i < N; i++) {
            if (indegree[i] == 0) {
                q.push(i);
            }
        }

        // Step 4: Kahn’s Algorithm
        int cnt = 0;
        while (!q.empty()) {
            int node = q.front();
            q.pop();
            cnt++;

            for (int it : adj[node]) {
                indegree[it]--;
                if (indegree[it] == 0) {
                    q.push(it);
                }
            }
        }

        // Step 5: Check cycle
        return cnt == N;   // true = possible, false = cycle
    }
};